### ISSUE TITLE
network_predict_gpu segfault from shared library but CPU works fine.

### ISSUE THREAD

**Issue by acidtonic** [AlexeyAB/darknet#3460]
I am running the latest darknet from master and receive a segfault when trying to use the API from C++. I can use the darknet binary with either CPU or GPU but when linking against the shared library only CPU works.

Gentoo Linux X86_64 with Intel Compiler

`#0  0x0000000006c85474 in __intel_avx_rep_memcpy ()
#1  0x00007ffff24f528e in network_predict_gpu (net=..., input=0x7fff00000001) at ./src/network_kernels.cu:488
#2  0x00007ffff24830f0 in network_predict () from /home/arkk/code/redacted/dep/darknet_AB/libdarknet.so
`
Inside network_predict_gpu the line it crashes on is "state.input = net.input_state_gpu;"

Using the cuda-gdb debugger I could print the net object coming into the network_predict_gpu function... not sure if it helps but here is that output...

`print net
$12 = {n = 107,	batch = 1, seen = 0x7fa9140, t = 0x0, epoch = 0, subdivisions = 1, layers = 0x7ffff7ef4010, output = 0x7fff73622e00, policy = STEPS, learning_rate = 0.00100000005,
  learning_rate_min = 9.99999975e-06, learning_rate_max = 0, batches_per_cycle = 1000, batches_cycle_mult = 2, momentum = 0.899999976, decay = 0.000500000024, gamma = 0, scale = 0,
  power = 4, time_steps = 1, step = 0, max_batches = 4000, seq_scales = 0x7fa9260, scales = 0x7fa9240, steps = 0x7fa9220, num_steps = 2, burn_in = 1000, cudnn_half = 0, adam = 0, B1 = 0,
  B2 = 0, eps = 0, inputs = 519168, outputs = 56784, truths = 0, notruth = 0, h = 416, w = 416,	c = 3, max_crop = 832, min_crop = 416, max_ratio = 0, min_ratio = 0, center = 0, flip = 1,
  blur = 0, mixup = 0, angle = 0, aspect = 1, exposure = 1.5, saturation = 1.5,	hue = 0.100000001, random = 0, track = 0, augment_speed = 2, sequential_subdivisions = 1,
  init_sequential_subdivisions = 1, current_subdivision = 0, try_fix_nan = 0, gpu_index = 0, hierarchy = 0x0, input = 0x0, truth = 0x0,	delta = 0x0, workspace = 0x7fff0c000000, train = 0,
  index = 0, cost = 0x0, clip = 0, delta_gpu = 0x3f00000000a73853, output_gpu = 0x1a03f000000, input_state_gpu = 0x7fff000001a0, input_pinned_cpu = 0x7fff00000001,
  input_pinned_cpu_flag = -14896, input_gpu = 0x7ffff7844a60 <std::cout>, truth_gpu = 0x7ffff7844a60 <std::cout>, input16_gpu = 0x2, output16_gpu = 0x7ffff7844a60 <std::cout>,
  max_input16_size = 0x7ffff7844a60 <std::cout>, max_output16_size = 0x7ffff7844a60 <std::cout>, wait_stream = 133838048}
`

I am using the basic example for doing detection in C++ and pointing at a cfg/weight/names/input file. I can run the detector with the darknet binary using the exact same cfg/weight/names/input and it works so I can only assume it has to do with the shared library or my usage of it.

Here is a quick dump of my latest try. I have toyed with using opencv for the image loading rather than load_image_color but had the same result with both.

`int darknet_main(int argc, char** argv) {

    static char *cfg_file = const_cast<char *>("db/train_arkk/arkk.cfg");
    static char *weight_file = const_cast<char *>("db/train_arkk/backup/arkk_step2_last.weights");
    static char *names_file = const_cast<char *>("db/train_arkk/arkk.names");   
    static char *input = const_cast<char *>("test_image.jpg");

    float thresh = 0.5;
    float hier_thresh = 0.5;

    size_t classes = 2;
    //char **labels = get_labels(names_file);  -> function missing, my cfg has 2 classes.
    //while (labels[classes] != nullptr) {
    //    classes++;
    //}

    network *net = load_network(cfg_file, weight_file, 0); //Tried with 1 for last arg, no difference
    // In case of testing (predicting a class), set batch number to 1, exact the way it needs to be set in *.cfg file
    //set_batch_network(net, 1);  (this was missing, I edited the cfg to have batch=1 assuming that did the same thing)

    cv::Mat frameDet = cv::imread ("test_image.jpg");
    cv::Mat resized;
    frameDet.convertTo(resized, CV_32FC3, 1.0/255.0);
    cv::cvtColor(resized, resized, cv::COLOR_BGR2RGB);
    cv::resize(resized, resized, cv::Size(net->w, net->h), 0, 0, cv::INTER_CUBIC);
    std::vector<cv::Mat> resizedChannels(3);
    split(resized, resizedChannels);
    vconcat(resizedChannels, resized);

    double time = what_time_is_it_now();
    network_predict(*net, resized.ptr<float>()); -> CRASH

    // other tries for loading image and calling predict -> same result
    //image im = load_image_color(input, 0, 0);
    //image sized = letterbox_image(im, net->w, net->h);
    //layer l = net->layers[net->n - 1];
    //float *frame_data = sized.data;
    //network_predict(*net, frame_data);

`

Any assistance or pointers would be helpful. I also noticed the line it crashes on has a commented out version above calling cuda_make_array. I switched the lines and recompiled and still it crashes in exactly the same way. It was worth a shot. Help!

Thanks.

**Comment by AlexeyAB** [comment]
Try to debug more.

At least, try to insert different printf("..."); after each line here to catch place with error: https://github.com/AlexeyAB/darknet/blob/9e9b2c493630a8ebd05d8732f0eb6b12a078bb88/src/network_kernels.cu#L478-L497

----

Can you successfully use this shared library from Python by using these examples?

* https://github.com/AlexeyAB/darknet/blob/master/darknet.py

* https://github.com/AlexeyAB/darknet/blob/master/darknet_video.py

**Comment by acidtonic** [comment]
Thank you for the reply, as I indicated in my report the line that fails is 487. All the values look fine to me. I will look into using this shared lib with python for sanity and report back.
-------- Original message --------From: Alexey <notifications@github.com> Date: 6/20/19  6:33 PM  (GMT-05:00) To: AlexeyAB/darknet <darknet@noreply.github.com> Cc: Zach <zdavisk@gmail.com>, Author <author@noreply.github.com> Subject: Re: [AlexeyAB/darknet] network_predict_gpu segfault from shared library but CPU works fine. (#3460) Try to debug more.
At least, try to insert different printf("..."); after each line here to catch place with error: https://github.com/AlexeyAB/darknet/blob/9e9b2c493630a8ebd05d8732f0eb6b12a078bb88/src/network_kernels.cu#L478-L497

Can you successfully use this shared library from Python by using these examples?


https://github.com/AlexeyAB/darknet/blob/master/darknet.py


https://github.com/AlexeyAB/darknet/blob/master/darknet_video.py



—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/AlexeyAB/darknet/issues/3460?email_source=notifications\u0026email_token=ABAB2Q5MNOY5BADTVBZMW5DP3QASPA5CNFSM4HZ77PGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYG5LLY#issuecomment-504223151",
"url": "https://github.com/AlexeyAB/darknet/issues/3460?email_source=notifications\u0026email_token=ABAB2Q5MNOY5BADTVBZMW5DP3QASPA5CNFSM4HZ77PGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYG5LLY#issuecomment-504223151",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by acidtonic** [comment]
Well I was hoping to find my library broken but after editing the paths to my configs into the python script it works just fine. The detections are accurate.

So python can use the shared library but my example code above cannot. Next I'll try to breakpoint the working code and see what parameters are different if any coming into network_predict_gpu().

I also turned off AVX and the only difference is the backtrace when it crashes is a memcpy rather than the intel AVX which leads me to believe it's a memory access or stack overflow issue from first glance.

Any other ideas or sample C++ code I could try? Really appreciate it.

**Comment by acidtonic** [comment]
Still investigating, not sure what else to look into. Valgrind reports this....  

==21134== Use of uninitialised value of size 8
==21134==    at 0x91F26F4: __intel_avx_rep_memcpy (in /home/arkk/intel/compilers_and_libraries_2019.4.243/linux/compiler/lib/intel64_lin/libintlc.so.5)
==21134==    by 0xE2C128D: network_predict_gpu (network_kernels.cu:488)
==21134==    by 0xE24F0EF: network_predict (in /home/arkk/code/[redacted]/dep/darknet_AB/libdarknet.so)
==21134==    by 0xA9BB4C: darknet_main(int, char**) (main.cpp:721)
==21134==    by 0xA9CFBA: main (main.cpp:98)
==21134== 
==21134== Invalid write of size 8
==21134==    at 0x91F26F4: __intel_avx_rep_memcpy (in /home/arkk/intel/compilers_and_libraries_2019.4.243/linux/compiler/lib/intel64_lin/libintlc.so.5)
==21134==    by 0xE2C128D: network_predict_gpu (network_kernels.cu:488)
==21134==    by 0xE24F0EF: network_predict (in /home/arkk/code/[redacted]/dep/darknet_AB/libdarknet.so)
==21134==    by 0xA9BB4C: darknet_main(int, char**) (main.cpp:721)
==21134==    by 0xA9CFBA: main (main.cpp:98)
==21134==  Address 0x1 is not stack'd, malloc'd or (recently) free'd
==21134== 
==21134== 
==21134== Process terminating with default action of signal 11 (SIGSEGV)
==21134==  Access not within mapped region at address 0x1
==21134==    at 0x91F26F4: __intel_avx_rep_memcpy (in /home/arkk/intel/compilers_and_libraries_2019.4.243/linux/compiler/lib/intel64_lin/libintlc.so.5)
==21134==    by 0xE2C128D: network_predict_gpu (network_kernels.cu:488)
==21134==    by 0xE24F0EF: network_predict (in /home/arkk/code/[redacted]/dep/darknet_AB/libdarknet.so)
==21134==    by 0xA9BB4C: darknet_main(int, char**) (main.cpp:721)
==21134==    by 0xA9CFBA: main (main.cpp:98)

**Comment by AlexeyAB** [comment]
It seems the error is there: https://github.com/AlexeyAB/darknet/blob/9e9b2c493630a8ebd05d8732f0eb6b12a078bb88/src/network_kernels.cu#L488

----

Try to comment these lines: https://github.com/AlexeyAB/darknet/blob/9e9b2c493630a8ebd05d8732f0eb6b12a078bb88/src/parser.c#L979-L983

And add this line:
`net.input_pinned_cpu = (float*)calloc(size, sizeof(float));`

**Comment by acidtonic** [comment]
Still crashes at the same place, changed to __memmove_avx_unaligned_erms() but otherwise still at the same line.

input_state_gpu is 0x7fff000001a0
print input
$13 = (float *) 0x5c460000
print size
$14 = 519168

Appreciate your help!

**Comment by acidtonic** [comment]
Just wanted to report that if I take the yolo console dll file and copy it into my project I can use it without crashing so something is definitely odd here.

In my case I'll just use the C++ API which for whatever reason doesn't crash like the C API does. Appreciate your help on this. I'd be happy to assist debugging this further if you are interested.

**Comment by hardik22317** [comment]
@AlexeyAB  I am also facing the same issue, getting a segmentation fault at the same line in net.input_pinned_cpu. So, I just want to ask that is there any update on this issue?
